jina/peapods/pods/__init__.py

```diff
@@ -409,6 +483,34 @@ def num_peas(self) -> int:
     def __eq__(self, other: 'BasePod'):
         return self.num_peas == other.num_peas and self.name == other.name
 
+    def _enter_pea(self, pea: 'BasePea') -> None:
+        self.peas.append(pea)
+        self.enter_context(pea)
+
+    def _activate(self):
+        # order is good. Activate from tail to head
+        for pea in reversed(self.peas):
+            if pea.args.socket_in == SocketType.DEALER_CONNECT:
+                pea.runtime.activate()
+
+        self._activated = True
+
+    def deactivate(self):
+        """Force Dealer Peas to send CANCEL messages to their ROUTERs
+
+        .. note:
+            This is to make sure Head of Replica always sends CANCEL in a blocking manner.
+            Like this we are sure that in CANCEL has been properly handled by the time the Head of CompoundPod
+            receives any Request and so can't send it to the wrong replica.
+
+        .. # noqa: DAR201
+        """
+        # order is good. Deactivate from head to tail
```

Comment:
I am not sure. Do I understand it correctly, that we currently do:
```
CANCEL HEAD
CANCEL Pea1
CANCEL Pea2
CANCEL TAIL
TERMINATE HEAD
TERMINATE Pea1
TERMINATE Pea2
TERMINATE TAIL
```
? If so, I believe we should do
```
CANCEL HEAD
TERMINATE HEAD
CANCEL Pea1
TERMINATE Pea1
CANCEL Pea2
TERMINATE Pea2
CANCEL TAIL
TERMINATE TAIL
```

Otherwise we cancel Pea1 before we are sure, that all messaged in Head are processed, non?

Reply:
I think it does not happen, but it may make sense. 
It is not such a trivial change. I would try in another PR.

Reply:
Then convince me, why it does not happen. Otherwise, this PR just shifts the problem to higher-load cases.

Reply:
It totally makes sense, actually. And the code will look much cleaner if deactivate is part of the closing routine of the Pea.

Reply:
I will do another iteration here to add unittests to the ExitFiFo and testing all this with Docker images.

This totally makes sense, I let you guys decide if i prefer to unblock the depending ticket or wait for this change to be made.

Reply:
Let's merge and unblock the depending PR and in parallel fix this deactivate/close orchestration issue.